file/src/main/scala/akka/stream/alpakka/file/scaladsl/LogRotatorSink.scala

```diff
@@ -8,163 +8,197 @@ import java.nio.file.{OpenOption, Path, StandardOpenOption}
+    )
+
+  /**
+   * Sink directing the incoming `ByteString`s to a new `Sink` created bye `sinkFactory` whenever `generatorFunction returns a new value.
+   *
+   * @param functionGeneratorFunction creates a function that triggers rotation by returning a value
+   * @param sinkFactory creates sinks for `ByteString`s from the value returned by `functionGeneratorFunction`
+   * @tparam C criterion type (for files a `Path`)
+   * @tparam R result type in materialized futures of `sinkFactory`
+   **/
+  def withSinkFactory[C, R](
+      functionGeneratorFunction: () => ByteString => Option[C],
+      sinkFactory: C => Sink[ByteString, Future[R]]
+  ): Sink[ByteString, Future[Done]] =
+    Sink.fromGraph(new LogRotatorSink[C, R](functionGeneratorFunction, sinkFactory))
 }
 
-final private[scaladsl] class LogRotatorSink(functionGeneratorFunction: () => ByteString => Option[Path],
-                                             fileOpenOptions: Set[OpenOption])
+/**
+ * "Log Rotator Sink" graph stage
+ *
+ * @param functionGeneratorFunction creates a function that triggers rotation by returning a value
+ * @param sinkFactory creates sinks for `ByteString`s from the value returned by `functionGeneratorFunction`
+ * @tparam C criterion type (for files a `Path`)
+ * @tparam R result type in materialized futures of `sinkFactory`
+ */
+final private class LogRotatorSink[C, R](functionGeneratorFunction: () => ByteString => Option[C],
+                                         sinkFactory: C => Sink[ByteString, Future[R]])
     extends GraphStageWithMaterializedValue[SinkShape[ByteString], Future[Done]] {
 
   val in = Inlet[ByteString]("FRotator.in")
   override val shape = SinkShape.of(in)
 
   override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
     val promise = Promise[Done]()
-    val logic = new GraphStageLogic(shape) {
-      val pathGeneratorFunction: ByteString => Option[Path] = functionGeneratorFunction()
-      var sourceOut: SubSourceOutlet[ByteString] = _
-      var fileSinkCompleted: Seq[Future[IOResult]] = Seq.empty
-      val decider =
-        inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
-
-      def failThisStage(ex: Throwable): Unit =
-        if (!promise.isCompleted) {
-          if (sourceOut != null) {
-            sourceOut.fail(ex)
-          }
-          cancel(in)
-          promise.failure(ex)
+    val decider =
+      inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    val logic = new Logic(promise, decider)
+    (logic, promise.future)
+  }
+
+  private final class Logic(promise: Promise[Done], decider: Decider) extends GraphStageLogic(shape) {
+    val pathGeneratorFunction: ByteString => Option[C] = functionGeneratorFunction()
+    var sourceOut: SubSourceOutlet[ByteString] = _
+    var fileSinkCompleted: immutable.Seq[Future[R]] = immutable.Seq.empty
```

Comment:
This now holds materialized values of generic sinks, therefor it should be renamed. `materializedValues`?